### PR TITLE
CTB: Trigger integration

### PR DESCRIPTION
**Pull request by MRiganSUSX**
Implementation of the CTB signals into trigger.

There is a new plugin `CTBTriggerCandidateMaker` based on the `TimingTriggerCandidateMaker`. The plugin is 'standalone' (not part of the TP flow) and connects to the MLT (and to TC Buffer). 
This plugin receives HSIEvents and converts them to TCs. New TC types have been added to reflect this (pybinds included).
The creation of the TCs works in a loop as it is possible to receive a signal with multiple bits on, each bit corresponding to a (to-be-made) TC of a different type. 
There is a map of signal bits to TC types in the header file.
There is an error thrown if the received bit is not known (not in the map).
There is functional opmon for the hsi events and TC objects. 

Because there are now many new TC types, several other wrappers had to be extended. This includes the `bitsets` in the MLT as well as the `trigger_type` in the TriggerDecision message. This is also propagated all the way to the TriggerRecordHeaderData. As a consequence the TRHD is now bigger (!) and padding had to be adjusted to accommodate for this.

!!! IMPORTANTLY, the version of both TriggerCandidateData and TriggerRecordHeaderData were increased to reflect these big changes. !!!
-

The CTB plugin has 3 configurable parameters:
	"ctb_prescale": 1,                 <- we can prescale the events
	"ctb_time_before": 1000,     <- extension to the readout window before hsievent timestamp
	"ctb_time_after": 1000,        <- extension to the readout window after hsievent timestamp

This passes the integration tests: minimal_system_quick_test.py, readout_type_scan.py, fake_data_producer_test.py.

From custom test, I was able to use fake_hsi to get HSI events of different signals.
These are picked by the CTB:
```
[CTB] Converting HSI event, signal: 8880680
```
Bits are extracted:
```
BITS: 00000000100001111000001000101000
```
and appropriate TCs are created:
```
this bit: 3
TC type: 13
```
These are received at MLT:
```
 Got TC of type 13, timestamp 79554162131545930, start/end 79554162131544930/79554162131546930
```
As was the case before, the trigger_type inside the HDF5 file corresponds to the trigger_type sent by the MLT.

### CODE REVIEW

**Comment by MRiganSUSX** [comment]
Accompanying PRs:
https://github.com/DUNE-DAQ/trgdataformats/pull/16
https://github.com/DUNE-DAQ/daqdataformats/pull/53
https://github.com/DUNE-DAQ/daqconf/pull/441
https://github.com/DUNE-DAQ/fddaqconf/pull/30

**Review by bieryAtFnal — APPROVED** [review]
My testing so far looks good.

One note is that we will need to change hdf5libs/scripts/hdf5_dump.py to take into account the larger size for TriggerRecordHeader::trigger_type, but I want to ask people about versioning inside that script, and I don't want to hold up the bulk of the changes in the meantime.

File: plugins/CTBTriggerCandidateMaker.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+      if (m_HLT_TC_map.count(i)) {
+        TLOG_DEBUG(TLVL_DEBUG_ALL) << "[CTB] TC type: " << static_cast<int>(m_HLT_TC_map[i]);
+
+        triggeralgs::TriggerCandidate candidate;
+        candidate.time_candidate = data.timestamp;
+        candidate.time_start = data.timestamp - m_time_before;
+        candidate.time_end = data.timestamp + m_time_after;
+        // candidate.detid = 1;
+        candidate.detid = data.header;
```
Comment (ArturSztuc) [inline plugins/CTBTriggerCandidateMaker.cpp:65]: I think this is probably the best way of dealing with "detid" here... `HSIEvent::header` tells us which device the event came from

File: plugins/CTBTriggerCandidateMaker.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+CTBTriggerCandidateMaker::HSIEventToTriggerCandidate(const dfmessages::HSIEvent& data)
+{
+  TLOG_DEBUG(TLVL_DEBUG_MEDIUM) << "[CTB] Converting HSI event, signal: " << data.signal_map;
+
+  std::vector<triggeralgs::TriggerCandidate> candidates;
+  std::bitset<32> bits(data.signal_map);
+  TLOG_DEBUG(TLVL_DEBUG_HIGH) << "[CTB] BITS: " << bits;
+
+  for (size_t i = 0; i < bits.size(); ++i) {
+    if (bits.test(i)) {
```
Comment (ArturSztuc) [inline plugins/CTBTriggerCandidateMaker.cpp:53]: Using guard clauses instead, e.g. 
```
if (!bits.test(i)) {
    continue;
}
```

would mean not having large nested `if`s. More here: https://refactoring.guru/replace-nested-conditional-with-guard-clauses

File: plugins/ModuleLevelTrigger.hpp
Comment (ArturSztuc) [inline plugins/ModuleLevelTrigger.hpp:1]: Wonder if the number of bits could be saved somewhere and used throughout, but I guess we won't be changing it too often...

**Review by ArturSztuc — APPROVED** [review]
LGTM. I have some comments, but nothing to hold up the merge.

**Comment by ArturSztuc** [comment]
Testing the CTB logic to be discussed at today's v4 meeting

**Pull request merged**